### PR TITLE
add TCP checksums for ipv6 -> ipv4 and a bunch of other stuff

### PR DESCRIPTION
**Pull request by siwiutki**

### CODE REVIEW

File: bpf/nat64.c [hunk 1]
```diff
@@ -119,20 +119,31 @@ int nat64(struct __sk_buff* skb)
...
 	ip.check = (__u16)~sum4;                // sum4 cannot be zero, so this is never 0xFFFF
 
+	__u64 diff = 0;
+
+	// calculate checksum difference before any helpers that modify
+	// packet's data are called, because verifier will invalidate
+	// all packet pointers
+	if (ip.protocol == IPPROTO_TCP) {
+		// cover both src and dst address change, buffer for dst addr field is
+		// right after src addr both for IPv4 and IPv6 header
```
Comment (aojea) [inline bpf/nat64.c]: just add a bit more of info so people can understand the solution, as per offline conversation 

https://www.rfc-editor.org/rfc/rfc2460.html#section-8.1
http://www.tcpipguide.com/free/t_TCPChecksumCalculationandtheTCPPseudoHeader-2.htm

File: README.md [hunk 2]
```diff
@@ -147,11 +157,11 @@ This is far to be complete, features and suggestions are welcome:
 - [ ] Testing, testing, ....
 
 
- 
+
```
Comment (aojea) [inline README.md:162]: Add a new section with contributors and add yourself to it

```
## Contributors

@siwiutki
```

File: bpf/nat64.c (hunk 1 above)
Comment (aojea) [reply bpf/nat64.c]: is just talk about the assumptions we are doing, omitting theother fields

File: bpf/nat64.c [hunk 3]
```diff
@@ -144,10 +155,14 @@ int nat64(struct __sk_buff* skb)
 		__u16 new_csum = 0;
 		bpf_skb_store_bytes(skb, UDP_CSUM_OFF, &new_csum, sizeof(new_csum), 0);
 	}
-	// TODO: recalculate TCP checksum since we can not depend on checksum neutrality
+	// recalculate TCP checksum since we can not depend on checksum neutrality
```
Comment (aojea) [inline bpf/nat64.c]: I prefer if we unify the solution and just recalculate it for both UDP and TCP

File: bpf/nat64.c [hunk 4]
```diff
@@ -133,6 +133,17 @@ int nat64(struct __sk_buff* skb)
 	sum4 = (sum4 & 0xFFFF) + (sum4 >> 16);  // collapse any potential carry into u16
 	ip.check = (__u16)~sum4;                // sum4 cannot be zero, so this is never 0xFFFF
 
+	__u64 diff = 0;
+
+	// calculate checksum difference before any helpers that modify
+	// packet's data are called, because verifier will invalidate
+	// all packet pointers
+	if (ip.protocol == IPPROTO_TCP) {
```
Comment (aojea) [inline bpf/nat64.c]: if TCP or UDP

**Comment by aojea** [comment]
almost there, some minor comments

File: hack/clean-leaked.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+function exec_kind_node {
+    ${CONTAINERD} exec -it "$@"
+}
+
+
+for kind_node in `kubectl get nodes | tail -n +2 | awk '{ print $1 }'`; do
+    # fetch iface with default route
+    GWIF="`exec_kind_node ${kind_node} ip route | grep default | awk '{ print $5 }'`"
+
+    # clean iptables rules
```
Comment (aojea) [inline hack/clean-leaked.sh:24]: I've added this logic to the golang program, is it not cleaning it?

the program should delete everything on exit

File: bpf/nat64.c (hunk 3 above)
Comment (siwiutki) [reply bpf/nat64.c]: Changed it so the checksum is calculated for both.

File: bpf/nat64.c (hunk 4 above)
Comment (siwiutki) [reply bpf/nat64.c]: Done.

File: bpf/nat64.c (hunk 1 above)
Comment (siwiutki) [reply bpf/nat64.c]: Added a rather lengthy explanation in comment.

**Comment by siwiutki** [comment]
weird thing - tcpdump is showing that checksum is bad, but I fiddled with checksum a bit, setting invalid values on purpose, and if it's indeed bad, UDP request does not come through, and same is for TCP. might be a signal that something is still wrong, but uh, requests work, so not sure what's up with that.

image

**Comment by aojea** [comment]
> weird thing - tcpdump is showing that checksum is bad

that is normal https://www.wireshark.org/faq.html#_why_am_i_seeing_lots_of_packets_with_incorrect_tcp_checksums

If the traffic works end to end then the checksum is right, previously the receiving part discarded the packet due to checksum incorrect

File: bpf/nat64.c [hunk 6]
```diff
@@ -43,127 +43,174 @@
...
+			// summing all bytes, zeroes does not matter),
+			// meaning we only need to run bpf_csum_diff over IP
+			// addresses instead of the entire pseudo-header.
+			//
+			// Last neat piece of info that makes it a one-liner is that both
+			// ipv6hdr and iphdr structs have src and dst addresses
+			// next to each other in memory. That means we can
+			// calculate checksum difference with one bpf_csum_diff
+			// call using 2 * size of IP address.
+			l4_csum_diff = bpf_csum_diff((void *)&(ip6->saddr), 2*sizeof(struct in6_addr), (void *)&(ip4.saddr), 2*sizeof(__u32), 0);
```
Comment (aojea) [inline bpf/nat64.c:164]: this is super smart and neat, great

File: bpf/nat64.c [hunk 7]
```diff
@@ -43,127 +43,174 @@
...
 	// believe the explicit check is required to keep the in kernel ebpf verifier happy.
-	if (data + l2_header_size + sizeof(struct iphdr) > data_end)
+	if (data + sizeof(struct ethhdr) + sizeof(struct iphdr) > data_end)
 		return TC_ACT_SHOT;
 
-	struct ethhdr* new_eth = data;
-	// Copy over the updated ethernet header
-	*new_eth = eth2;
+	// Copy over the old ethernet header with updated ethertype.
+	ret = bpf_skb_store_bytes(skb, 0, &old_eth, sizeof(struct ethhdr), 0);
```
Comment (aojea) [inline bpf/nat64.c:202]: reading the docs https://man7.org/linux/man-pages/man7/bpf-helpers.7.html it seems this is safer than the previous approach that manipulated directly the skb

File: bpf/nat64.c [hunk 8]
```diff
@@ -277,6 +256,38 @@ static __always_inline int nat46(struct __sk_buff *skb)
...
+			// If IPv4/UDP checksum is 0 then fallback to clatd so it can calculate the
+			// checksum. Otherwise the network or more likely the NAT64 gateway might
+			// drop the packet because in most cases IPv6/UDP packets with a zero checksum
+			// are invalid. See RFC 6935.
+			if (!uh->check) {
+				bpf_printk("NAT46 packet forwarded: UDP checksum is 0");
+				return TC_ACT_OK;
+			}
+			// If checksum is not 0, recalculate with bpf_csum_diff.
+			// Fallthrough to the next case.
```
Comment (aojea) [inline bpf/nat64.c:278]: this is an artifact from my copy paste from the android implementation, 0 checksum in UDP for IPv4 is ok, so we need to recalculate the checksum, This seems to rely on another component called clatd, but we are not using it, so just recalculate if is possible

**Comment by aojea** [comment]
This looks great, thanks for fixing it, just a follow up https://github.com/aojea/nat64/pull/1#discussion_r1518819883

**Pull request merged**